File: VB.Net/PIT_Magic/PitData.vb
Imports Microsoft.VisualBasic.CompilerServices
Imports System
Imports System.Collections.Generic
Imports System.IO
Imports System.Runtime.Serialization.Formatters.Binary

Namespace PIT_Magic
    <Serializable> _
    Public Class PitData
        ' Methods
        Public Sub AddEntry(ByVal entry As PitEntry)
            entry.Index = Me._entryCount
            Me._entries.Add(entry)
            Me._entryCount += 1
        End Sub

        Public Sub Clear()
            Me._entryCount = 0
            Me._dummyData1 = New Byte() { 0, 0, 0, 0 }
            Me._dummyData2 = New Byte() { 0, 0, 0, 0 }
            Me._dummyData3 = New Byte() { 0, 0, 0, 0 }
            Me._dummyData4 = New Byte() { 0, 0, 0, 0 }
            Me._dummyData5 = New Byte() { 0, 0, 0, 0 }
            Me._entries.Clear
        End Sub

        Public Shared Function Clone(ByVal pData As PitData) As PitData
            Using stream As MemoryStream = New MemoryStream
                Dim formatter As New BinaryFormatter
                formatter.Serialize(stream, pData)
                stream.Position = 0
                Return DirectCast(formatter.Deserialize(stream), PitData)
            End Using
        End Function

        Private Function CompareByteArray(ByVal byteArray1 As Byte(), ByVal byteArray2 As Byte()) As Boolean
            If (Not byteArray1 Is byteArray2) Then
                If ((byteArray1 Is Nothing) OrElse (byteArray2 Is Nothing)) Then
                    Return False
                End If
                If (byteArray1.Length <> byteArray2.Length) Then
                    Return False
                End If
                Dim num2 As Integer = (byteArray1.Length - 1)
                Dim i As Integer = 0
                Do While (i <= num2)
                    If (byteArray1(i) <> byteArray2(i)) Then
                        Return False
                    End If
                    i += 1
                Loop
            End If
            Return True
        End Function

        Public Function FindEntry(ByVal partitionName As String) As PitEntry
            Dim num As Integer = 0
            Do While (num < Me._entries.Count)
                Dim entry2 As PitEntry = Me._entries.Item(num)
                If String.Equals(partitionName, entry2.PartitionName) Then
                    Return entry2
                End If
            Loop
            Return Nothing
        End Function

        Public Function GetEntry(ByVal index As Integer) As PitEntry
            Return Me._entries.Item(index)
        End Function

        Public Function Matches(ByVal otherPitData As PitData) As Boolean
            If Not ((((((Me._entryCount = otherPitData._entryCount) And Me.CompareByteArray(Me._dummyData1, otherPitData._dummyData1)) And Me.CompareByteArray(Me._dummyData2, otherPitData._dummyData2)) And Me.CompareByteArray(Me._dummyData3, otherPitData._dummyData3)) And Me.CompareByteArray(Me._dummyData4, otherPitData._dummyData4)) And Me.CompareByteArray(Me._dummyData5, otherPitData._dummyData5)) Then
                Return False
            End If
            Dim i As Integer
            For i = 0 To Me._entryCount - 1
                Dim entry As PitEntry = Me._entries.Item(i)
                Dim otherPitEntry As PitEntry = otherPitData._entries.Item(i)
                If Conversions.ToBoolean(Operators.NotObject(entry.Matches(otherPitEntry))) Then
                    Return False
                End If
            Next i
            Return True
        End Function

        Public Function ReadPITFile(ByVal inputStream As PitInputStream) As Boolean
            Dim buffer As Byte() = New Byte(4  - 1) {}
            Dim buffer2 As Byte() = New Byte(&H20  - 1) {}
            If (inputStream.ReadInt <> &H12349876) Then
                Return False
            End If
            Me._entries.Clear
            Me._entryCount = inputStream.ReadInt
            Me._entries.Capacity = Me._entryCount
            Me._dummyData1 = inputStream.ReadBytes(4)
            Me._dummyData2 = inputStream.ReadBytes(4)
            Me._dummyData3 = inputStream.ReadBytes(4)
            Me._dummyData4 = inputStream.ReadBytes(4)
            Me._dummyData5 = inputStream.ReadBytes(4)
            Dim i As Integer
            For i = 0 To Me._entryCount - 1
                Dim item As New PitEntry With { _
                    .Index = i _
                }
                item.SetEntryMemAddr(inputStream.GetMemoryAddress)
                Dim num2 As Integer = inputStream.ReadInt
                item.BinaryType = DirectCast(num2, EntryBinaryType)
                num2 = inputStream.ReadInt
                item.DeviceType = DirectCast(num2, EntryDeviceType)
                num2 = inputStream.ReadInt
                item.Identifier = num2
                num2 = inputStream.ReadInt
                item.Attribute = DirectCast(num2, EntryAttribute)
                num2 = inputStream.ReadInt
                item.UpdateAttribute = DirectCast(num2, EntryUpdateAttribute)
                num2 = inputStream.ReadInt
                item.BlockSize = num2
                num2 = inputStream.ReadInt
                item.BlockCount = num2
                num2 = inputStream.ReadInt
                item.FileOffset = num2
                num2 = inputStream.ReadInt
                item.FileSize = num2
                item.PartitionName = inputStream.ReadString(buffer2, &H20)
                item.FlashFileName = inputStream.ReadString(buffer2, &H20)
                item.FotaFileName = inputStream.ReadString(buffer2, &H20)
                Me._entries.Add(item)
            Next i
            Return True
        End Function

        Public Sub RemoveEntry(ByVal index As Integer)
            Dim enumerator As Enumerator(Of PitEntry)
            Me._entries.RemoveAt(index)
            Me._entryCount -= 1
            Dim num As Integer = 0
            Try 
                enumerator = Me._entries.GetEnumerator
                Do While enumerator.MoveNext
                    Dim current As PitEntry = enumerator.Current
                    current.Index = num
                    num += 1
                Loop
            Finally
                enumerator.Dispose
            End Try
        End Sub

        Public Sub SetEntry(ByVal index As Integer, ByVal newPitEntry As PitEntry)
            Me._entries.RemoveAt(index)
            Me._entries.Insert(index, newPitEntry)
        End Sub

        Public Sub WritePITFile(ByVal outputStream As PitOutputStream)
            outputStream.WriteInt(&H12349876)
            outputStream.WriteInt(Me._entryCount)
            outputStream.WriteBytes(Me._dummyData1)
            outputStream.WriteBytes(Me._dummyData2)
            outputStream.WriteBytes(Me._dummyData3)
            outputStream.WriteBytes(Me._dummyData4)
            outputStream.WriteBytes(Me._dummyData5)
            Dim i As Integer
            For i = 0 To Me._entryCount - 1
                Dim entry As PitEntry = Me._entries.Item(i)
                outputStream.WriteInt(CInt(entry.BinaryType))
                outputStream.WriteInt(CInt(entry.DeviceType))
                outputStream.WriteInt(entry.Identifier)
                outputStream.WriteInt(CInt(entry.Attribute))
                outputStream.WriteInt(CInt(entry.UpdateAttribute))
                outputStream.WriteInt(entry.BlockSize)
                outputStream.WriteInt(entry.BlockCount)
                outputStream.WriteInt(entry.FileOffset)
                outputStream.WriteInt(entry.FileSize)
                outputStream.WriteString(entry.PartitionName, &H20)
                outputStream.WriteString(entry.FlashFileName, &H20)
                outputStream.WriteString(entry.FotaFileName, &H20)
            Next i
        End Sub


        ' Properties
        Public Property DummyData1 As Byte()
            Get
                Return Me._dummyData1
            End Get
            Set(ByVal newValue As Byte())
                Me._dummyData1 = newValue
            End Set
        End Property

        Public Property DummyData2 As Byte()
            Get
                Return Me._dummyData2
            End Get
            Set(ByVal newValue As Byte())
                Me._dummyData2 = newValue
            End Set
        End Property

        Public Property DummyData3 As Byte()
            Get
                Return Me._dummyData3
            End Get
            Set(ByVal newValue As Byte())
                Me._dummyData3 = newValue
            End Set
        End Property

        Public Property DummyData4 As Byte()
            Get
                Return Me._dummyData4
            End Get
            Set(ByVal newValue As Byte())
                Me._dummyData4 = newValue
            End Set
        End Property

        Public Property DummyData5 As Byte()
            Get
                Return Me._dummyData5
            End Get
            Set(ByVal newValue As Byte())
                Me._dummyData5 = newValue
            End Set
        End Property

        Public ReadOnly Property Entries As PitEntry()
            Get
                Return Me._entries.ToArray
            End Get
        End Property

        Public ReadOnly Property EntryCount As Integer
            Get
                Return Me._entryCount
            End Get
        End Property


        ' Fields
        Private _dummyData1 As Byte() = New Byte() { 0, 0, 0, 0 }
        Private _dummyData2 As Byte() = New Byte() { 0, 0, 0, 0 }
        Private _dummyData3 As Byte() = New Byte() { 0, 0, 0, 0 }
        Private _dummyData4 As Byte() = New Byte() { 0, 0, 0, 0 }
        Private _dummyData5 As Byte() = New Byte() { 0, 0, 0, 0 }
        Private _entries As List(Of PitEntry) = New List(Of PitEntry)
        Private _entryCount As Integer
        Public Const DummyDataBlockLength As Integer = 4
        Public Const HeaderMagic As Integer = &H12349876
        Public Const HeaderSize As Integer = &H1C
        Public Const StrMaxLength As Integer = &H20
    End Class
End Namespace

File: C#/PIT_Magic/PitData.cs
namespace PIT_Magic
{
    using Microsoft.VisualBasic.CompilerServices;
    using System;
    using System.Collections.Generic;
    using System.IO;
    using System.Runtime.Serialization.Formatters.Binary;

    [Serializable]
    public class PitData
    {
        private byte[] _dummyData1 = new byte[] { 0, 0, 0, 0 };
        private byte[] _dummyData2 = new byte[] { 0, 0, 0, 0 };
        private byte[] _dummyData3 = new byte[] { 0, 0, 0, 0 };
        private byte[] _dummyData4 = new byte[] { 0, 0, 0, 0 };
        private byte[] _dummyData5 = new byte[] { 0, 0, 0, 0 };
        private List<PitEntry> _entries = new List<PitEntry>();
        private int _entryCount;
        public const int DummyDataBlockLength = 4;
        public const int HeaderMagic = 0x12349876;
        public const int HeaderSize = 0x1c;
        public const int StrMaxLength = 0x20;

        public void AddEntry(PitEntry entry)
        {
            entry.Index = this._entryCount;
            this._entries.Add(entry);
            this._entryCount++;
        }

        public void Clear()
        {
            this._entryCount = 0;
            this._dummyData1 = new byte[] { 0, 0, 0, 0 };
            this._dummyData2 = new byte[] { 0, 0, 0, 0 };
            this._dummyData3 = new byte[] { 0, 0, 0, 0 };
            this._dummyData4 = new byte[] { 0, 0, 0, 0 };
            this._dummyData5 = new byte[] { 0, 0, 0, 0 };
            this._entries.Clear();
        }

        public static PitData Clone(PitData pData)
        {
            using (MemoryStream stream = new MemoryStream())
            {
                BinaryFormatter formatter = new BinaryFormatter();
                formatter.Serialize(stream, pData);
                stream.Position = 0L;
                return (PitData) formatter.Deserialize(stream);
            }
        }

        private bool CompareByteArray(byte[] byteArray1, byte[] byteArray2)
        {
            if (byteArray1 != byteArray2)
            {
                if ((byteArray1 == null) || (byteArray2 == null))
                {
                    return false;
                }
                if (byteArray1.Length != byteArray2.Length)
                {
                    return false;
                }
                int num2 = byteArray1.Length - 1;
                for (int i = 0; i <= num2; i++)
                {
                    if (byteArray1[i] != byteArray2[i])
                    {
                        return false;
                    }
                }
            }
            return true;
        }

        public PitEntry FindEntry(string partitionName)
        {
            int num = 0;
            while (num < this._entries.Count)
            {
                PitEntry entry2 = this._entries[num];
                if (string.Equals(partitionName, entry2.PartitionName))
                {
                    return entry2;
                }
            }
            return null;
        }

        public PitEntry GetEntry(int index) => 
            this._entries[index];

        public bool Matches(PitData otherPitData)
        {
            if (!((((((this._entryCount == otherPitData._entryCount) & this.CompareByteArray(this._dummyData1, otherPitData._dummyData1)) & this.CompareByteArray(this._dummyData2, otherPitData._dummyData2)) & this.CompareByteArray(this._dummyData3, otherPitData._dummyData3)) & this.CompareByteArray(this._dummyData4, otherPitData._dummyData4)) & this.CompareByteArray(this._dummyData5, otherPitData._dummyData5)))
            {
                return false;
            }
            for (int i = 0; i < this._entryCount; i++)
            {
                PitEntry entry = this._entries[i];
                PitEntry otherPitEntry = otherPitData._entries[i];
                if (Conversions.ToBoolean(Operators.NotObject(entry.Matches(otherPitEntry))))
                {
                    return false;
                }
            }
            return true;
        }

        public bool ReadPITFile(PitInputStream inputStream)
        {
            byte[] buffer = new byte[4];
            byte[] buffer2 = new byte[0x20];
            if (inputStream.ReadInt() != 0x12349876)
            {
                return false;
            }
            this._entries.Clear();
            this._entryCount = inputStream.ReadInt();
            this._entries.Capacity = this._entryCount;
            this._dummyData1 = inputStream.ReadBytes(4);
            this._dummyData2 = inputStream.ReadBytes(4);
            this._dummyData3 = inputStream.ReadBytes(4);
            this._dummyData4 = inputStream.ReadBytes(4);
            this._dummyData5 = inputStream.ReadBytes(4);
            for (int i = 0; i < this._entryCount; i++)
            {
                PitEntry item = new PitEntry {
                    Index = i
                };
                item.SetEntryMemAddr(inputStream.GetMemoryAddress());
                int num2 = inputStream.ReadInt();
                item.BinaryType = (PitEntry.EntryBinaryType) num2;
                num2 = inputStream.ReadInt();
                item.DeviceType = (PitEntry.EntryDeviceType) num2;
                num2 = inputStream.ReadInt();
                item.Identifier = num2;
                num2 = inputStream.ReadInt();
                item.Attribute = (PitEntry.EntryAttribute) num2;
                num2 = inputStream.ReadInt();
                item.UpdateAttribute = (PitEntry.EntryUpdateAttribute) num2;
                num2 = inputStream.ReadInt();
                item.BlockSize = num2;
                num2 = inputStream.ReadInt();
                item.BlockCount = num2;
                num2 = inputStream.ReadInt();
                item.FileOffset = num2;
                num2 = inputStream.ReadInt();
                item.FileSize = num2;
                item.PartitionName = inputStream.ReadString(buffer2, 0x20);
                item.FlashFileName = inputStream.ReadString(buffer2, 0x20);
                item.FotaFileName = inputStream.ReadString(buffer2, 0x20);
                this._entries.Add(item);
            }
            return true;
        }

        public void RemoveEntry(int index)
        {
            List<PitEntry>.Enumerator enumerator;
            this._entries.RemoveAt(index);
            this._entryCount--;
            int num = 0;
            try
            {
                enumerator = this._entries.GetEnumerator();
                while (enumerator.MoveNext())
                {
                    PitEntry current = enumerator.Current;
                    current.Index = num;
                    num++;
                }
            }
            finally
            {
                enumerator.Dispose();
            }
        }

        public void SetEntry(int index, PitEntry newPitEntry)
        {
            this._entries.RemoveAt(index);
            this._entries.Insert(index, newPitEntry);
        }

        public void WritePITFile(PitOutputStream outputStream)
        {
            outputStream.WriteInt(0x12349876);
            outputStream.WriteInt(this._entryCount);
            outputStream.WriteBytes(this._dummyData1);
            outputStream.WriteBytes(this._dummyData2);
            outputStream.WriteBytes(this._dummyData3);
            outputStream.WriteBytes(this._dummyData4);
            outputStream.WriteBytes(this._dummyData5);
            for (int i = 0; i < this._entryCount; i++)
            {
                PitEntry entry = this._entries[i];
                outputStream.WriteInt((int) entry.BinaryType);
                outputStream.WriteInt((int) entry.DeviceType);
                outputStream.WriteInt(entry.Identifier);
                outputStream.WriteInt((int) entry.Attribute);
                outputStream.WriteInt((int) entry.UpdateAttribute);
                outputStream.WriteInt(entry.BlockSize);
                outputStream.WriteInt(entry.BlockCount);
                outputStream.WriteInt(entry.FileOffset);
                outputStream.WriteInt(entry.FileSize);
                outputStream.WriteString(entry.PartitionName, 0x20);
                outputStream.WriteString(entry.FlashFileName, 0x20);
                outputStream.WriteString(entry.FotaFileName, 0x20);
            }
        }

        public byte[] DummyData1
        {
            get => 
                this._dummyData1;
            set
            {
                this._dummyData1 = value;
            }
        }

        public byte[] DummyData2
        {
            get => 
                this._dummyData2;
            set
            {
                this._dummyData2 = value;
            }
        }

        public byte[] DummyData3
        {
            get => 
                this._dummyData3;
            set
            {
                this._dummyData3 = value;
            }
        }

        public byte[] DummyData4
        {
            get => 
                this._dummyData4;
            set
            {
                this._dummyData4 = value;
            }
        }

        public byte[] DummyData5
        {
            get => 
                this._dummyData5;
            set
            {
                this._dummyData5 = value;
            }
        }

        public PitEntry[] Entries =>
            this._entries.ToArray();

        public int EntryCount =>
            this._entryCount;
    }
}

